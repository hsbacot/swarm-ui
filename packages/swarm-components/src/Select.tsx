import * as React from 'react';
import { ArrowDown } from '@meetup/swarm-icons/lib/components/solid';

type Props = React.ElementConfig<HTMLSelectElement> & {
	/**
	 * Whether the select menu is disabled.
	 */
	disabled?: boolean,
	/**
	 * Whether the select menu has errored.
	 */
	error?: string,
	/**
	 * Unique identifier for select tag.
	 */
	id?: string,
	/**
	 * Name of select menu form field.
	 */
	name: string,
	/**
	 * Copy to be used for a required select field.
	 */
	requiredText?: string,
};

const Select = (props: Props) => {
	const { disabled, error, id, name, requiredText, ...other } = props;

	const selectState = disabled ? 'disabled' : error ? 'error' : 'default';

	return (
		<div data-swarm-select={selectState}>
			<div data-swarm-select-wrapper>
				<select
					name={name}
					id={id || name}
					required={Boolean(requiredText)}
					disabled={disabled}
					{...other}
				/>

				<span data-swarm-select-arrow="arrow-down">
					<ArrowDown />
				</span>
			</div>
		</div>
	);
};

export default Select;
