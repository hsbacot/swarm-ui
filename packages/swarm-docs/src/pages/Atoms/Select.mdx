import { Select, FieldLabel } from  '@meetup/swarm-components';
import SelectExample from '../../components/examples/select';
import PropsTable from '../../components/propstable';

## Select
---
A wrapper for an html `<select>` element with an optional `<label>` tag and helper text.

<br/>
<br/>

### Snippet
---

Use this interactive snippet to try the component:
<br/>

```.jsx
<div style={{width: 200}}>
    <FieldLabel htmlFor="swarm-select" >Swarm Select</FieldLabel>
    <Select
        value="A"
        id="swarm-select"
        name="swarm-select"
        onChange={(e) => alert(e.target.value)}
    >
        <option value="A">
            A
        </option>
        <option value="B">
            B
        </option>
        <option value="C">
            C
        </option>
        <option value="D">
            D
        </option>
    </Select>
</div>
```
<br/>
<br/>

### Props
---
<br/>
<PropsTable propMetaData={Select.__docgenInfo} />


<br/>
<br/>

### a11y
---
<br/>

The following keyboard interactions are used to control the select component.
You can also filter by typing letters, but this behavior varies by browser.
Some will filter as you type, like autocomplete. Others will only sort by first letter.
E.g., in a list of US States, hitting A then R may take you to Arizona, or it may
take you to Alabama and then Rhode Island.
<br/>

**Keyboard interactions:**
* **↑/↓**       ... navigate between menu options
* **Spacebar**  ... expand

<br/>
See: <a href="https://webaim.org/techniques/keyboard/">https://webaim.org/techniques/keyboard/</a>

<br/>
<br/>

### Examples
---
<br/>

<SelectExample />
