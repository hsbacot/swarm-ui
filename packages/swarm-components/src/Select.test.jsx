import * as React from 'react';
import { shallow } from 'enzyme';
import { Select } from './Select';

describe('Button', () => {
	let count = 0;

	const getSelectMarkup = props => (
		<Select key={++count} {...props}>
			<option value="geoffrey">Geoffrey</option>
			<option value="drhorse">Doctor Horse, MD Junior</option>
			<option value="chompyhorse">Mister Chompy</option>
		</Select>
	);

	const testCases = [
		['Default', getSelectMarkup({ name: 'select' })],
		['Disabled', getSelectMarkup({ name: 'select', disabled: true })],
		['Error', getSelectMarkup({ name: 'select', error: true })],
		['With Id', getSelectMarkup({ name: 'select', id: 'select' })],
		['Required Text', getSelectMarkup({ name: 'select', requireText: 'Must fill!' })],
	];

	test.each(testCases)('Snapshot: %s', (description, element) => {
		expect(shallow(element)).toMatchSnapshot();
	});
});
