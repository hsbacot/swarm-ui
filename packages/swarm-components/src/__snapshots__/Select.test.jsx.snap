// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Snapshot: Default 1`] = `
<div
  data-swarm-select="default"
>
  <div
    data-swarm-select-wrapper={true}
  >
    <select
      id="select"
      name="select"
      required={false}
    >
      <option
        value="geoffrey"
      >
        Geoffrey
      </option>
      <option
        value="drhorse"
      >
        Doctor Horse, MD Junior
      </option>
      <option
        value="chompyhorse"
      >
        Mister Chompy
      </option>
    </select>
    <span
      data-swarm-select-arrow="arrow-down"
    >
      <ArrowDown />
    </span>
  </div>
</div>
`;

exports[`Button Snapshot: Disabled 1`] = `
<div
  data-swarm-select="disabled"
>
  <div
    data-swarm-select-wrapper={true}
  >
    <select
      disabled={true}
      id="select"
      name="select"
      required={false}
    >
      <option
        value="geoffrey"
      >
        Geoffrey
      </option>
      <option
        value="drhorse"
      >
        Doctor Horse, MD Junior
      </option>
      <option
        value="chompyhorse"
      >
        Mister Chompy
      </option>
    </select>
    <span
      data-swarm-select-arrow="arrow-down"
    >
      <ArrowDown />
    </span>
  </div>
</div>
`;

exports[`Button Snapshot: Error 1`] = `
<div
  data-swarm-select="error"
>
  <div
    data-swarm-select-wrapper={true}
  >
    <select
      id="select"
      name="select"
      required={false}
    >
      <option
        value="geoffrey"
      >
        Geoffrey
      </option>
      <option
        value="drhorse"
      >
        Doctor Horse, MD Junior
      </option>
      <option
        value="chompyhorse"
      >
        Mister Chompy
      </option>
    </select>
    <span
      data-swarm-select-arrow="arrow-down"
    >
      <ArrowDown />
    </span>
  </div>
</div>
`;

exports[`Button Snapshot: Required Text 1`] = `
<div
  data-swarm-select="default"
>
  <div
    data-swarm-select-wrapper={true}
  >
    <select
      id="select"
      name="select"
      requireText="Must fill!"
      required={false}
    >
      <option
        value="geoffrey"
      >
        Geoffrey
      </option>
      <option
        value="drhorse"
      >
        Doctor Horse, MD Junior
      </option>
      <option
        value="chompyhorse"
      >
        Mister Chompy
      </option>
    </select>
    <span
      data-swarm-select-arrow="arrow-down"
    >
      <ArrowDown />
    </span>
  </div>
</div>
`;

exports[`Button Snapshot: With Id 1`] = `
<div
  data-swarm-select="default"
>
  <div
    data-swarm-select-wrapper={true}
  >
    <select
      id="select"
      name="select"
      required={false}
    >
      <option
        value="geoffrey"
      >
        Geoffrey
      </option>
      <option
        value="drhorse"
      >
        Doctor Horse, MD Junior
      </option>
      <option
        value="chompyhorse"
      >
        Mister Chompy
      </option>
    </select>
    <span
      data-swarm-select-arrow="arrow-down"
    >
      <ArrowDown />
    </span>
  </div>
</div>
`;
